src/services/code-index/managed/watcher.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * File watcher for managed codebase indexing
+ *
+ * This module provides functions for watching file changes and incrementally
+ * updating the index. Changes are debounced and batched for efficiency.
+ */
+
+import * as vscode from "vscode"
+import * as path from "path"
+import { execSync } from "child_process"
+import { indexFile, handleFileDeleted } from "./scanner"
+import { ManagedIndexingConfig, FileChangeEvent } from "./types"
+import { scannerExtensions } from "../shared/supported-extensions"
+import { MANAGED_FILE_WATCH_DEBOUNCE_MS } from "../constants"
+import { logger } from "../../../utils/logging"
+import { TelemetryService } from "@roo-code/telemetry"
+import { TelemetryEventName } from "@roo-code/types"
+
+/**
+ * Creates and initializes a file watcher for managed indexing
+ *
+ * The watcher:
+ * - Monitors file create, change, and delete events
+ * - Debounces rapid changes (500ms default)
+ * - Filters by supported file extensions
+ * - Updates the index incrementally
+ *
+ * @param config Managed indexing configuration
+ * @param context VSCode extension context
+ * @param onFilesChanged Optional callback when files are processed
+ * @returns Disposable watcher instance
+ */
+export function createFileWatcher(
+	config: ManagedIndexingConfig,
+	context: vscode.ExtensionContext,
+	onFilesChanged?: (events: FileChangeEvent[]) => void,
+): vscode.Disposable {
+	// Create file system watcher for all files
+	const watcher = vscode.workspace.createFileSystemWatcher("**/*")
```

Comment:
Hmm, this pattern has caused a few performance issues in the past in other parts of the code. We have it [here](https://github.com/Kilo-Org/kilocode/blob/main/src/integrations/workspace/WorkspaceTracker.ts#L43) and it tends to blow up my CPU when I do something like an `npm install` since it sees all those files getting created/destroyed. It's not fixed there either, but just something to mention!